pandas/plotting/_matplotlib/core.py

```diff
@@ -234,6 +249,52 @@ def _validate_color_args(self):
                             "pass 'style' without a color symbol"
                         )
 
+    @staticmethod
+    def _create_iter_data_given_by(
+        data: ABCDataFrame, by: Optional[list]
+    ) -> Union[ABCDataFrame, Dict[str, Union[ABCDataFrame, ABCSeries]]]:
+        """
+        Create data for iteration given `by` is assigned or not, and it is only
+        used in both hist and boxplot.
+
+        If `by` is assigned, return a dictionary of DataFrames in which the key of
+        dictionary is the values in groups.
+        If `by` is not assigned, return input as is, and this preserves current
+        status of iter_data.
+
+        Parameters
+        ----------
+        data: reformatted grouped data from `_compute_plot_data` method
+        by: list or None, value assigned to `by`.
+
+        Returns
+        -------
+        iter_data: DataFrame or Dictionary of DataFrames
+
+        Examples
+        --------
+        If `by` is assigned:
+
+        >>> tuples = [('h1', 'a'), ('h1', 'b'), ('h2', 'a'), ('h2', 'b')]
+        >>> mi = MultiIndex.from_tuples(tuples)
+        >>> value = [[1, 3, np.nan, np.nan],
+        ...          [3, 4, np.nan, np.nan], [np.nan, np.nan, 5, 6]]
+        >>> data = DataFrame(value, columns=mi) # doctest: +SKIP
+        >>> _create_iter_data_given_by(data, by=["col1"]) # doctest: +SKIP
+        {'h1': DataFrame({'a': [1, 3, np.nan], 'b': [3, 4, np.nan]}),
+         'h2': DataFrame({'a': [np.nan, np.nan, 5], 'b': [np.nan, np.nan, 6]})}
+        """
+        if not by:
```

Comment:
Hmm we do something like this generically in groupby already (see `_iterate_slices`) which I guess begs the question, is there a reason for implementing this as a keyword rather than through the existing groupby plotting API? 

Apologies in advance if that was already asked

Reply:
Thanks for taking a look. And here the reason to implement it as keyword is to avoid using groupby plotting api, because the implementation of groupby plotting is different than general plotting api, thus actually providing slightly different output figure. Also, I don't think existing groupby plotting api supports `hist`? it only supports `boxplot`, right?

Reply:
So what do you imagine the difference between `df.hist(..., by="col")` vs `df.groupby("col").hist()`? Not sure if there is a problem here or not, just wary of terminology

Reply:
yeah, actually `df.plot(kind='hist', by='col')` and `df.hist( by='col')` are following two different code paths, and the output will be slightly different. One is through `_grouped_hist` (what we use for `df.hist()`) and the other is through `_plot` in `HistPlot`, and also the arguments those two have are slightly different, e.g. `subplots` not supported in `df.hist` API etc. And i believe the ultimate goal would be deprecating the `hist_frame`, `hist_series` etc, and fully use plotting API.

sorry, I made some graphs for comparison, but are hidden because we've got quite many conversations in total ^^ 
https://github.com/pandas-dev/pandas/pull/28373#issuecomment-570814191

and also this was observed quite long ago: https://github.com/pandas-dev/pandas/issues/6279#issuecomment-35021300

The issue I am trying to implement here is to support `df.plot(kind='hist', by=)` and the way how `plot` works here is through the `iter_data` and generates plots one by one, so that what we need is to make data iterrable to be able to be directly used in plotting function, and some logics/legend etc are changed inside plotting API through `generate` method, therefore, what i think needs to be done is to mimic how `grouped_hist` works and to tweak grouped data for `_plot` in `HistPlot`.

And we can NOT directly use existing group API, otherwise, in this case, there will be two axes generated (one is empty from calling plot API) and one from existing group API which will get those plots.

I am sorry this PR has been here for quite long, and there have been quite many conversations. I wish I explained clearly, and if you have further questions, feel free to keep asking. @WillAyd